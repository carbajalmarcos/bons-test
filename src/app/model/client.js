import { ApolloClient, InMemoryCache, HttpLink } from "@apollo/client";
const httpLink = new HttpLink({
  uri: "http://game.bons.me/graphql",
});

const cache = new InMemoryCache({ addTypename: false });

const client = new ApolloClient({
  link: httpLink, // concat errorLink to authlink and httplink, errorLink will catch errors
  cache,
});

export default client;
