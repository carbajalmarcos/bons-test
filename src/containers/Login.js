import React, { forwardRef } from "react";
import LogingForm from "../components/LoginForm";
const Login = () => {
  return <LogingForm />;
};

export default Login;
