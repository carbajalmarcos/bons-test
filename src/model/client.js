import { ApolloClient, InMemoryCache } from "@apollo/client";
const cache = new InMemoryCache({ addTypename: false });
const client = new ApolloClient({
  uri: "http://game.bons.me/graphql",
  cache,
});

export default client;
