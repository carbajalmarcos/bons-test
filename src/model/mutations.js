import { gql } from "@apollo/client";
export const CREATE_GAME = gql`
  mutation createGame($name: String!) {
    createGame(input: { name: $name }) {
      id
      currentTurn
      maxTurns
      turnsLeft
      monster {
        hp
        id
        maxHp
        name
        image
        shield
      }
      player {
        hp
        id
        maxHp
        name
        shield
        cards {
          effect
          id
          value
        }
      }
    }
  }
`;
const NEXT_TURN = gql`
  mutation nextTurn($gameId: String!, $cardId: String) {
    nextTurn(input: { gameId: $gameId, cardId: $cardId }) {
      game {
        currentTurn
        id
        maxTurns
        monster {
          hp
          id
          image
          maxHp
          name
          shield
        }
        player {
          cards {
            effect
            id
            value
          }
          hp
          id
          maxHp
          name
          shield
        }
        turnsLeft
      }
      monsterEffect {
        effect
        value
      }
    }
  }
`;
