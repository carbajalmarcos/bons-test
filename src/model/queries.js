import { gql } from "@apollo/client";

export const GET_GAME = gql`
  query getGame {
    game @client
  }
`;

export const GET_MONSTER_EFFECT = gql`
  query getMonsterEffect {
    monsterEffect @client
  }
`;
