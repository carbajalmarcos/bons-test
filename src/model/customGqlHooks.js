import { CREATE_GAME, NEXT_TURN } from "./mutations";
import { GET_GAME } from "./queries";
import { useQuery, useMutation } from "@apollo/client";

export const useCreateGame = () => {
  return useMutation(CREATE_GAME, {
    update(cache, { data: { createGame } }) {
      cache.writeQuery({
        query: GET_GAME,
        data: { game: createGame, monsterEffect: {} },
      });
    },
  });
};
export const useNextTurn = () => {
  return useMutation(NEXT_TURN, {
    update(cache, { data: { nextTurn } }) {
      console.log(nextTurn);
      cache.writeQuery({
        query: GET_GAME,
        data: { ...nextTurn },
      });
    },
  });
};
export const useGetGame = () => {
  return useQuery(GET_GAME);
};
