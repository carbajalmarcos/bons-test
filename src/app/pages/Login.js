import React, { useEffect } from "react";
import LogingForm from "../components/LoginForm";
import { useCreateGame } from "../../model/customGqlHooks";
import { useHistory } from "react-router-dom";
import Modal from "../components/Modal";
import Spinner from "../components/Spinner";

const Login = () => {
  const [createGame, { data, loading, error }] = useCreateGame();
  const history = useHistory();

  const handleCreateGame = (name) => {
    createGame({ variables: { name } });
  };

  const handleOnclose = () => {
    history.push("/");
  };

  useEffect(() => {
    if (data && !loading) {
      history.push("gameboard");
    }
  }, [data, loading]);

  if (error) {
    return (
      <Modal
        content={<p>{error.message}</p>}
        onClose={handleOnclose}
        data-test="modal"
      />
    );
  }
  return (
    <>
      {loading && <Spinner data-test="login-spinner" />}
      <LogingForm createGame={handleCreateGame} />;
    </>
  );
};

export default Login;
