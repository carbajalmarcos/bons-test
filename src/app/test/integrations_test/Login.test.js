import React from "react";
import Enzyme, { mount } from "enzyme";
import EnzymeAdapter from "enzyme-adapter-react-16";
import { findByTestAttr } from "../testUtils";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import Login from "../../pages/Login";
import { MockedProvider } from "@apollo/react-testing";
import { createGameMock } from "../mocks";
import { CREATE_GAME } from "../../../model/mutations";
Enzyme.configure({
  adapter: new EnzymeAdapter(),
});
const setup = (mocks) => {
  return mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Router>
        <Switch>
          <Route exact path="/" component={Login} />
        </Switch>
      </Router>
    </MockedProvider>
  );
};

describe("some of login flow test ", () => {
  test("name empty", () => {
    const loginMutationMock = [
      {
        request: {
          query: CREATE_GAME,
          variables: {
            name: "marcos",
          },
        },
        result: () => {
          return {
            data: createGameMock,
          };
        },
      },
    ];

    const wrapper = setup(loginMutationMock);
    const input = findByTestAttr(wrapper, "login-name");
    const mockEvent = { target: { value: "" } };

    input.first().simulate("change", mockEvent);

    const element = findByTestAttr(wrapper, "login-button");
    element.first().simulate("click");

    // let spinner = findByTestAttr(wrapper, "login-spinner");
    // expect(spinner.type().name).toEqual("Spinner");
    expect(wrapper.children().debug()).not.toContain("Spinner");
  });

  test("loding test", async () => {
    const loginMutationMock = [
      {
        request: {
          query: CREATE_GAME,
          variables: {
            name: "marcos",
          },
        },
        result: () => {
          return {
            data: createGameMock,
          };
        },
      },
    ];

    const wrapper = setup(loginMutationMock);
    const input = findByTestAttr(wrapper, "login-name");
    const mockEvent = { target: { value: "marcos" } };

    input.first().simulate("change", mockEvent);

    const element = findByTestAttr(wrapper, "login-button");
    element.first().simulate("click");

    // let spinner = findByTestAttr(wrapper, "login-spinner");
    // expect(spinner.type().name).toEqual("Spinner");
    expect(wrapper.children().debug()).toContain("Spinner");
    await new Promise((resolver) => setTimeout(resolver, 0));
    wrapper.update();
    expect(wrapper.children().debug()).not.toContain("Spinner");
  });
});
