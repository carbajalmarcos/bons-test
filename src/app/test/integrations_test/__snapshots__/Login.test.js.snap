// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render loading state initially 1`] = `[Function]`;
